Upwind finite difference SBP solver

ranocha opened the pull request:
Implements a new solver type that uses the upwind SBP operators with flux vector splitting.

**Note: This should be merged using a regular merge commit, not squash-merged, to preserve the history of some modifications along the development process and give credits to everyone involved.**

## TODO
- [x] First round of clean up (#1276)
- [x] `FluxUpwind` as an alternative numerical flux parameterized by splitting (#1278)
- [x] First performance testing (https://github.com/trixi-framework/Trixi.jl/pull/1282)
- [x] Tutorial and NEWS.md entry (https://github.com/trixi-framework/Trixi.jl/pull/1283)
- Corresponding Trixi2Vtk.jl updates (https://github.com/trixi-framework/Trixi2Vtk.jl/pull/57)

CC @andrewwinters5000 @sloede

ranocha commented:
@sloede This should be ready for a review of the user-facing interface. To get started, you can also look at the brief tutorial I wrote in https://github.com/trixi-framework/Trixi.jl/pull/1283.

sloede reviewed (commented):
Overall this looks very good! Here's a brief review of the public facing parts.

ranocha commented on src/auxiliary/math.jl:
```diff
@@ -192,5 +192,27 @@ julia> min(2, 5, 1)
+Return `x` if `x` is positive, else zero. In other words, return
+`(x + abs(x)) / 2` for real numbers `x`.
+"""
+@inline function positive_part(x)
+  o = zero(x)
+  return ifelse(x > o, x, o)
+end
```
It's basically equivalent (up to the behavior in the presence of signed zeros) - due to our use of `max = @fastmath max` above in the same file.

sloede replied:
Thanks for the clarification /c

ranocha commented on src/equations/compressible_euler_2d.jl:
```diff
@@ -664,6 +664,304 @@ end
+"""
+@inline function splitting_steger_warming(u, orientation::Integer,
+                                          equations::CompressibleEulerEquations2D)
+  fm = splitting_steger_warming(u, Val{:minus}(), orientation, equations)
+  fp = splitting_steger_warming(u, Val{:plus}(),  orientation, equations)
```
`Val{:minus}()` and `Val{:plus}()` are basically named singletons. This does not allocate:
```julia
julia> trixi_include("examples/tree_3d_fdsbp/elixir_euler_taylor_green_vortex.jl", tspan=(0.0, 1.0))
[...]
 ────────────────────────────────────────────────────────────────────────────────────
              Trixi.jl                      Time                    Allocations      
                                   ───────────────────────   ────────────────────────
         Tot / % measured:              25.3s /  88.4%            129MiB /  46.9%    

 Section                   ncalls     time    %tot     avg     alloc    %tot      avg
 ────────────────────────────────────────────────────────────────────────────────────
 rhs!                         850    22.1s   98.9%  26.0ms   9.33KiB    0.0%    11.2B
   volume integral            850    18.1s   81.0%  21.3ms     0.00B    0.0%    0.00B
   surface integral           850    1.63s    7.3%  1.92ms     0.00B    0.0%    0.00B
   interface flux             850    1.02s    4.5%  1.20ms     0.00B    0.0%    0.00B
   prolong2interfaces         850    593ms    2.7%   698μs     0.00B    0.0%    0.00B
   Jacobian                   850    513ms    2.3%   604μs     0.00B    0.0%    0.00B
   reset ∂u/∂t                850    237ms    1.1%   278μs     0.00B    0.0%    0.00B
   ~rhs!~                     850   8.07ms    0.0%  9.49μs   9.33KiB    0.0%    11.2B
   prolong2boundaries         850    331μs    0.0%   389ns     0.00B    0.0%    0.00B
   boundary flux              850   81.5μs    0.0%  95.9ns     0.00B    0.0%    0.00B
   mortar flux                850   17.0μs    0.0%  20.0ns     0.00B    0.0%    0.00B
   source terms               850   13.7μs    0.0%  16.1ns     0.00B    0.0%    0.00B
   prolong2mortars            850   13.2μs    0.0%  15.6ns     0.00B    0.0%    0.00B
 analyze solution               2    150ms    0.7%  75.2ms    192KiB    0.3%  96.0KiB
 I/O                            3   98.7ms    0.4%  32.9ms   60.1MiB   99.7%  20.0MiB
   save solution                2   97.4ms    0.4%  48.7ms   60.0MiB   99.6%  30.0MiB
   ~I/O~                        3   1.27ms    0.0%   425μs   19.8KiB    0.0%  6.60KiB
   get element variables        2   10.3μs    0.0%  5.15μs   16.4KiB    0.0%  8.20KiB
   save mesh                    2    258ns    0.0%   129ns     0.00B    0.0%    0.00B
 ────────────────────────────────────────────────────────────────────────────────────
```

sloede replied:
OK, makes sense, thanks! /c

ranocha commented on src/equations/numerical_fluxes.jl:
```diff
@@ -318,4 +318,28 @@ end
+A numerical flux `f(u_left, u_right) = f⁺(u_left) + f⁻(u_right)` based on
+flux vector splitting.
+
+The [`SurfaceIntegralUpwind`](@ref) with a given `splitting` is analytically
+equivalent to the [`SurfaceIntegralStrongForm`](@ref) with `FluxUpwind(splitting)`
+as numerical flux.
```
The latter. I reformulated it as
```
The [`SurfaceIntegralUpwind`](@ref) with a given `splitting` is equivalent to
the [`SurfaceIntegralStrongForm`](@ref) with `FluxUpwind(splitting)`
as numerical flux (up to floating point differences).
```
and will push it soon (in a new PR bundling these changes).

sloede replied:
Looks good in #1285! /c

ranocha commented on src/equations/numerical_fluxes.jl:
```diff
@@ -318,4 +318,28 @@ end
+
+@inline function (numflux::FluxUpwind)(u_ll, u_rr, orientation::Int, equations)
+  @unpack splitting = numflux
+  fm = splitting(u_rr, Val{:minus}(), orientation, equations)
+  fp = splitting(u_ll, Val{:plus}(),  orientation, equations)
```
Because there is basically nothing to be gained (in terms of performance) and the current version is more explicit. In a combined call, it may not be clear whether `f_minus = f_minus(u_ll)`, `f_minus = f_minus(u_rr)`, or something depending on both `u_ll` and `u_rr`.

sloede replied:
No, makes absolutely sense as you did it! /c

ranocha commented on src/solvers/fdsbp_tree/fdsbp.jl:
```diff
@@ -0,0 +1,51 @@
+eachnode(D::AbstractDerivativeOperator) = Base.OneTo(nnodes(D))
+get_nodes(D::AbstractDerivativeOperator) = grid(D)
+
+# TODO: This is hack to enable the FDSBP solver to use the
```
Tracked in https://github.com/trixi-framework/Trixi.jl/issues/1284

sloede replied:
Thanks! /c

ranocha commented on src/equations/compressible_euler_1d.jl:
```diff
@@ -360,6 +360,266 @@ end
 
 
+"""
+    splitting_steger_warming(u, orientation::Integer,
```
:+1:

ranocha commented on src/equations/compressible_euler_2d.jl:
```diff
@@ -308,7 +308,7 @@ Should be used together with [`UnstructuredMesh2D`](@ref).
   if v_normal <= 0.0
     sound_speed = sqrt(equations.gamma * p_local / rho_local) # local sound speed
-    p_star = p_local * (1.0 + 0.5 * (equations.gamma - 1) * v_normal / sound_speed)^(2.0 * equations.gamma * equations.inv_gamma_minus_one)
+    p_star = p_local * (1.0 + 0.5 * (equations.gamma - 1) * v_normal / sound_speed)^(2 * equations.gamma * equations.inv_gamma_minus_one)
```
I just didn't see it back then :sweat_smile:

andrewwinters5000 replied:
This was something that Gregor and I discussed in the initial implementation. When to use literals like `1.0` versus `1` because in Fortran these are different. As a follow-up, is it okay to use `1` here because it is adding to other reals but in the if statement we need to use `v_normal <= 0.0` because the `0.0` stand alone?

ranocha replied:
In this case, both versions are equivalent - as long as we use `Float64` for our computations:
```julia
julia> foo(x) = 1 + x
foo (generic function with 1 method)

julia> bar(x) = 1.0 + x
bar (generic function with 1 method)

julia> @code_native syntax=:intel debuginfo=:none foo(1.0)
        .text
        .file   "foo"
        .section        .rodata.cst8,"aM",@progbits,8
        .p2align        3                               # -- Begin function julia_foo_175
.LCPI0_0:
        .quad   0x3ff0000000000000              # double 1
        .text
        .globl  julia_foo_175
        .p2align        4, 0x90
        .type   julia_foo_175,@function
julia_foo_175:                          # @julia_foo_175
        .cfi_startproc
# %bb.0:                                # %top
        movabs  rax, offset .LCPI0_0
        vaddsd  xmm0, xmm0, qword ptr [rax]
        ret
.Lfunc_end0:
        .size   julia_foo_175, .Lfunc_end0-julia_foo_175
        .cfi_endproc
                                        # -- End function
        .section        ".note.GNU-stack","",@progbits

julia> @code_native syntax=:intel debuginfo=:none bar(1.0)
        .text
        .file   "bar"
        .section        .rodata.cst8,"aM",@progbits,8
        .p2align        3                               # -- Begin function julia_bar_173
.LCPI0_0:
        .quad   0x3ff0000000000000              # double 1
        .text
        .globl  julia_bar_173
        .p2align        4, 0x90
        .type   julia_bar_173,@function
julia_bar_173:                          # @julia_bar_173
        .cfi_startproc
# %bb.0:                                # %top
        movabs  rax, offset .LCPI0_0
        vaddsd  xmm0, xmm0, qword ptr [rax]
        ret
.Lfunc_end0:
        .size   julia_bar_173, .Lfunc_end0-julia_bar_173
        .cfi_endproc
                                        # -- End function
        .section        ".note.GNU-stack","",@progbits
```
But `Float64` coefficients stand in the way of running Trixi.jl with `Float32`, so I remove them when I see them from time to time. See also https://docs.julialang.org/en/v1/manual/style-guide/#Avoid-using-floats-for-numeric-literals-in-generic-code-when-possible

andrewwinters5000 replied:
Aha, gotcha. Thanks for the explanation!

ranocha commented on src/solvers/fdsbp_tree/fdsbp.jl:
```diff
@@ -0,0 +1,51 @@
+
+
+# For dispatch
+const FDSBP = DG{Basis} where {Basis<:AbstractDerivativeOperator}
```
I added a docstring and introduced `FDSBP` as constructor. Thus, users will see `solver = FDSBP(...)` in our example elixirs, which may clarify the intents.

sloede replied:
This was great both from a usability and an aesthetic POV! /c

ranocha commented:
> Overall this looks very good! Here's a brief review of the public facing parts.

Thanks for your review, @sloede! I made a new PR addressing your comments (https://github.com/trixi-framework/Trixi.jl/pull/1285).

sloede commented:
> > Overall this looks very good! Here's a brief review of the public facing parts.
> 
> Thanks for your review, @sloede! I made a new PR addressing your comments (#1285).

Thanks! I looked through all your responses and #1285. Once the latter is merged, I think you are good to go. The only thing you might want to consider is adding the `FDSBP` solver to the table in https://trixi-framework.github.io/Trixi.jl/stable/overview/#Semidiscretizations, maybe with a footnote that it still experimental.

ranocha commented:
Thanks! This is a good suggestion. I noted it in the corresponding doc PR.

ranocha commented:
I will merge this when tests pass. Then, we can get a new release to test the corresponding setup in Trixi2Vtk.jl.

sloede commented:
🙀

ranocha commented:
> :scream_cat:

Not okay?

sloede commented:
> > 🙀
> 
> Not okay?

No, no, all good! Just crazy how fast everything went, and how we have a new solver 😊...
